pkg/backend/leader_elect.go

```diff
@@ -0,0 +1,227 @@
+package backend
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	klogv2 "k8s.io/klog/v2"
+
+	"github.com/Azure/azure-sdk-for-go/storage"
+
+	"github.com/khenidak/london/pkg/backend/consts"
+	filterutils "github.com/khenidak/london/pkg/backend/filter"
+	storageerrors "github.com/khenidak/london/pkg/backend/storageerrors"
+	"github.com/khenidak/london/pkg/types"
+)
+
+/* We are not doing a traditional leader-elect here. The primary difference being the caller
+is responsible for renewing the leader-elect-lease. We are not running a go-routine on behalf of caller and renewing it.
+// TODO: do we need a release api?
+
+*/
+
+type le struct {
+	lock sync.Mutex
+	// election name
+	electionName string
+	// my name is current process, host whatever competes for the election
+	myName string
+
+	s *store
+	e *storage.Entity
+}
+
+// creates new election based on storage table
+func (s *store) NewLeaderElection(electionName string, myName string) types.LeaderElect {
+	// note we are not testing names here. Thus user can create empty electionName and myName
+	return &le{
+		electionName: electionName,
+		myName:       myName,
+		s:            s,
+	}
+}
+
+func (l *le) ElectionName() string {
+	return l.electionName
+}
+
+func (l *le) MyName() string {
+	return l.myName
+}
+
+// elects current leader
+func (l *le) Elect(duration time.Duration) (bool, error) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	// first try to blindly insert a new entity
+	e := createElectionEntity(l, duration)
+	e.Table = l.s.t // assigns the table
+	batch := l.s.t.NewBatch()
+	batch.InsertEntity(e)
+	err := batch.ExecuteBatch()
+
+	// assuming that we are clear
+	if err == nil {
+		// optionally save entity
+		_ = l.ensureCurrentEntity()
+		klogv2.Infof("leader election:%v is currently owned by:%v", l.electionName, l.myName)
+		return true, nil
+	}
+
+	if !storageerrors.IsEntityAlreadyExists(err) {
+		// this is not an error we can deal with
+		return false, err
+	}
+
+	return l.renewTerm(duration)
+}
+
+func (l *le) CurrentHolder() (string, error) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	if err := l.ensureCurrentEntity(); err != nil {
+		return "", err
+	}
+
+	return l.e.Properties[consts.LeaderElectOwnerNameFieldName].(string), nil
+}
+
+func (l *le) RenewTerm(duration time.Duration) (bool, error) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	return l.renewTerm(duration)
+}
+func (l *le) renewTerm(duration time.Duration) (bool, error) {
```

Comment:
For both the cases after the check for current holder or not (lease expired) there is quite a bit of common code. Adding a comment here so we can tackle this during refactor